From b7641ff010c7708bd79dd816c4b98598064d0171 Mon Sep 17 00:00:00 2001
From: woot000 <woot000@woot000.com>
Date: Fri, 11 Jul 2025 19:16:34 -0500
Subject: [PATCH] CMake: Add ENABLE_EXTRA_EXAMPLES

Separate executables enabled with ENABLE_EXAMPLES that are not installed by default (i.e. not aomenc or aomdec) behind ENABLE_EXTRA_EXAMPLES, to reduce build times and save space by creating less executables
---
 CMakeLists.txt | 166 ++++++++++++++++++++++++++-----------------------
 1 file changed, 87 insertions(+), 79 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 48d92f6a4b..ef17a40400 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -457,18 +457,23 @@ if(CONFIG_AV1_DECODER AND ENABLE_EXAMPLES)
   add_executable(aomdec "${AOM_ROOT}/apps/aomdec.c"
                         $<TARGET_OBJECTS:aom_common_app_util>
                         $<TARGET_OBJECTS:aom_decoder_app_util>)
-  add_executable(decode_to_md5 "${AOM_ROOT}/examples/decode_to_md5.c"
-                               $<TARGET_OBJECTS:aom_common_app_util>
-                               $<TARGET_OBJECTS:aom_decoder_app_util>)
-  add_executable(decode_with_drops "${AOM_ROOT}/examples/decode_with_drops.c"
-                                   $<TARGET_OBJECTS:aom_common_app_util>
-                                   $<TARGET_OBJECTS:aom_decoder_app_util>)
-  add_executable(simple_decoder "${AOM_ROOT}/examples/simple_decoder.c"
-                                $<TARGET_OBJECTS:aom_common_app_util>
-                                $<TARGET_OBJECTS:aom_decoder_app_util>)
-  add_executable(scalable_decoder "${AOM_ROOT}/examples/scalable_decoder.c"
+
+  if(ENABLE_EXTRA_EXAMPLES)
+    add_executable(decode_to_md5 "${AOM_ROOT}/examples/decode_to_md5.c"
+                                 $<TARGET_OBJECTS:aom_common_app_util>
+                                 $<TARGET_OBJECTS:aom_decoder_app_util>)
+    add_executable(decode_with_drops "${AOM_ROOT}/examples/decode_with_drops.c"
+                                     $<TARGET_OBJECTS:aom_common_app_util>
+                                     $<TARGET_OBJECTS:aom_decoder_app_util>)
+    add_executable(simple_decoder "${AOM_ROOT}/examples/simple_decoder.c"
                                   $<TARGET_OBJECTS:aom_common_app_util>
                                   $<TARGET_OBJECTS:aom_decoder_app_util>)
+    add_executable(scalable_decoder "${AOM_ROOT}/examples/scalable_decoder.c"
+                                    $<TARGET_OBJECTS:aom_common_app_util>
+                                    $<TARGET_OBJECTS:aom_decoder_app_util>)
+    list(APPEND AOM_DECODER_EXAMPLE_TARGETS decode_to_md5 decode_with_drops
+               scalable_decoder simple_decoder)
+  endif()
 
   if(CONFIG_ANALYZER)
     add_executable(analyzer "${AOM_ROOT}/examples/analyzer.cc"
@@ -509,8 +514,7 @@ if(CONFIG_AV1_DECODER AND ENABLE_EXAMPLES)
   endif()
 
   # Maintain a list of decoder example targets.
-  list(APPEND AOM_DECODER_EXAMPLE_TARGETS aomdec decode_to_md5 decode_with_drops
-              scalable_decoder simple_decoder)
+  list(APPEND AOM_DECODER_EXAMPLE_TARGETS aomdec)
 
   # Add decoder examples to the app targets list.
   list(APPEND AOM_APP_TARGETS ${AOM_DECODER_EXAMPLE_TARGETS})
@@ -541,45 +545,47 @@ if(CONFIG_AV1_ENCODER)
                           $<TARGET_OBJECTS:aom_common_app_util>
                           $<TARGET_OBJECTS:aom_encoder_app_util>
                           $<TARGET_OBJECTS:aom_encoder_stats>)
-    add_executable(lossless_encoder "${AOM_ROOT}/examples/lossless_encoder.c"
+
+    if(ENABLE_EXTRA_EXAMPLES)
+      add_executable(lossless_encoder "${AOM_ROOT}/examples/lossless_encoder.c"
+                                      $<TARGET_OBJECTS:aom_common_app_util>
+                                      $<TARGET_OBJECTS:aom_encoder_app_util>)
+      add_executable(set_maps "${AOM_ROOT}/examples/set_maps.c"
+                              $<TARGET_OBJECTS:aom_common_app_util>
+                              $<TARGET_OBJECTS:aom_encoder_app_util>)
+      add_executable(simple_encoder "${AOM_ROOT}/examples/simple_encoder.c"
                                     $<TARGET_OBJECTS:aom_common_app_util>
                                     $<TARGET_OBJECTS:aom_encoder_app_util>)
-    add_executable(set_maps "${AOM_ROOT}/examples/set_maps.c"
-                            $<TARGET_OBJECTS:aom_common_app_util>
-                            $<TARGET_OBJECTS:aom_encoder_app_util>)
-    add_executable(simple_encoder "${AOM_ROOT}/examples/simple_encoder.c"
-                                  $<TARGET_OBJECTS:aom_common_app_util>
-                                  $<TARGET_OBJECTS:aom_encoder_app_util>)
-    add_executable(twopass_encoder "${AOM_ROOT}/examples/twopass_encoder.c"
+      add_executable(twopass_encoder "${AOM_ROOT}/examples/twopass_encoder.c"
+                                     $<TARGET_OBJECTS:aom_common_app_util>
+                                     $<TARGET_OBJECTS:aom_encoder_app_util>)
+      if(NOT BUILD_SHARED_LIBS AND NOT CONFIG_REALTIME_ONLY)
+        add_executable(noise_model "${AOM_ROOT}/examples/noise_model.c"
                                    $<TARGET_OBJECTS:aom_common_app_util>
                                    $<TARGET_OBJECTS:aom_encoder_app_util>)
-    if(NOT BUILD_SHARED_LIBS AND NOT CONFIG_REALTIME_ONLY)
-      add_executable(noise_model "${AOM_ROOT}/examples/noise_model.c"
-                                 $<TARGET_OBJECTS:aom_common_app_util>
-                                 $<TARGET_OBJECTS:aom_encoder_app_util>)
-      add_executable(photon_noise_table
-                     "${AOM_ROOT}/examples/photon_noise_table.c"
-                     $<TARGET_OBJECTS:aom_common_app_util>
-                     $<TARGET_OBJECTS:aom_encoder_app_util>)
+        add_executable(photon_noise_table
+                       "${AOM_ROOT}/examples/photon_noise_table.c"
+                       $<TARGET_OBJECTS:aom_common_app_util>
+                       $<TARGET_OBJECTS:aom_encoder_app_util>)
+        list(APPEND AOM_ENCODER_EXAMPLE_TARGETS noise_model photon_noise_table)
+      endif()
+      add_executable(scalable_encoder "${AOM_ROOT}/examples/scalable_encoder.c"
+                                      $<TARGET_OBJECTS:aom_common_app_util>
+                                      $<TARGET_OBJECTS:aom_encoder_app_util>)
+      add_executable(svc_encoder_rtc "${AOM_ROOT}/examples/svc_encoder_rtc.cc"
+                                     $<TARGET_OBJECTS:aom_common_app_util>
+                                     $<TARGET_OBJECTS:aom_encoder_app_util>)
+      target_link_libraries(svc_encoder_rtc ${AOM_LIB_LINK_TYPE} aom_av1_rc)
+    list(APPEND AOM_ENCODER_EXAMPLE_TARGETS lossless_encoder set_maps
+                simple_encoder scalable_encoder svc_encoder_rtc twopass_encoder)
     endif()
-    add_executable(scalable_encoder "${AOM_ROOT}/examples/scalable_encoder.c"
-                                    $<TARGET_OBJECTS:aom_common_app_util>
-                                    $<TARGET_OBJECTS:aom_encoder_app_util>)
-    add_executable(svc_encoder_rtc "${AOM_ROOT}/examples/svc_encoder_rtc.cc"
-                                   $<TARGET_OBJECTS:aom_common_app_util>
-                                   $<TARGET_OBJECTS:aom_encoder_app_util>)
-    target_link_libraries(svc_encoder_rtc ${AOM_LIB_LINK_TYPE} aom_av1_rc)
 
     # Maintain a list of encoder example targets.
-    list(APPEND AOM_ENCODER_EXAMPLE_TARGETS aomenc lossless_encoder set_maps
-                simple_encoder scalable_encoder svc_encoder_rtc twopass_encoder)
-    if(NOT BUILD_SHARED_LIBS AND NOT CONFIG_REALTIME_ONLY)
-      list(APPEND AOM_ENCODER_EXAMPLE_TARGETS noise_model photon_noise_table)
-    endif()
+    list(APPEND AOM_ENCODER_EXAMPLE_TARGETS aomenc)
 
   endif()
 
-  if(ENABLE_TOOLS)
+  if(ENABLE_EXTRA_EXAMPLES AND ENABLE_TOOLS)
     if(CONFIG_ENTROPY_STATS AND NOT BUILD_SHARED_LIBS)
 
       # TODO(tomfinegan): Sort out why a simple link command with
@@ -737,7 +743,7 @@ if(ENABLE_EXAMPLES)
 endif()
 
 if(ENABLE_TOOLS)
-  if(CONFIG_AV1_DECODER)
+  if(CONFIG_AV1_DECODER AND ENABLE_EXTRA_EXAMPLES)
     add_executable(dump_obu "${AOM_ROOT}/tools/dump_obu.cc"
                             "${AOM_ROOT}/tools/obu_parser.cc"
                             "${AOM_ROOT}/tools/obu_parser.h"
@@ -756,47 +762,49 @@ if(ENABLE_TOOLS)
   endif()
 endif()
 
-if(ENABLE_EXAMPLES AND CONFIG_AV1_DECODER AND CONFIG_AV1_ENCODER)
-  add_executable(aom_cx_set_ref "${AOM_ROOT}/examples/aom_cx_set_ref.c"
-                                $<TARGET_OBJECTS:aom_common_app_util>
-                                $<TARGET_OBJECTS:aom_encoder_app_util>)
-  list(APPEND AOM_EXAMPLE_TARGETS aom_cx_set_ref)
-  list(APPEND AOM_APP_TARGETS aom_cx_set_ref)
-endif()
+if(ENABLE_EXTRA_EXAMPLES)
+  if(ENABLE_EXAMPLES AND CONFIG_AV1_DECODER AND CONFIG_AV1_ENCODER)
+    add_executable(aom_cx_set_ref "${AOM_ROOT}/examples/aom_cx_set_ref.c"
+                                  $<TARGET_OBJECTS:aom_common_app_util>
+                                  $<TARGET_OBJECTS:aom_encoder_app_util>)
+    list(APPEND AOM_EXAMPLE_TARGETS aom_cx_set_ref)
+    list(APPEND AOM_APP_TARGETS aom_cx_set_ref)
+  endif()
 
-if(ENABLE_EXAMPLES AND CONFIG_AV1_ENCODER)
-  add_executable(lightfield_encoder "${AOM_ROOT}/examples/lightfield_encoder.c"
-                                    $<TARGET_OBJECTS:aom_common_app_util>
-                                    $<TARGET_OBJECTS:aom_encoder_app_util>)
-  list(APPEND AOM_EXAMPLE_TARGETS lightfield_encoder)
-  list(APPEND AOM_APP_TARGETS lightfield_encoder)
-endif()
+  if(ENABLE_EXAMPLES AND CONFIG_AV1_ENCODER)
+    add_executable(lightfield_encoder "${AOM_ROOT}/examples/lightfield_encoder.c"
+                                      $<TARGET_OBJECTS:aom_common_app_util>
+                                      $<TARGET_OBJECTS:aom_encoder_app_util>)
+    list(APPEND AOM_EXAMPLE_TARGETS lightfield_encoder)
+    list(APPEND AOM_APP_TARGETS lightfield_encoder)
+  endif()
 
-if(ENABLE_EXAMPLES AND CONFIG_AV1_DECODER)
-  add_executable(lightfield_tile_list_decoder
-                 "${AOM_ROOT}/examples/lightfield_tile_list_decoder.c"
-                 $<TARGET_OBJECTS:aom_common_app_util>
-                 $<TARGET_OBJECTS:aom_decoder_app_util>)
-  list(APPEND AOM_EXAMPLE_TARGETS lightfield_tile_list_decoder)
-  list(APPEND AOM_APP_TARGETS lightfield_tile_list_decoder)
-endif()
+  if(ENABLE_EXAMPLES AND CONFIG_AV1_DECODER)
+    add_executable(lightfield_tile_list_decoder
+                   "${AOM_ROOT}/examples/lightfield_tile_list_decoder.c"
+                   $<TARGET_OBJECTS:aom_common_app_util>
+                   $<TARGET_OBJECTS:aom_decoder_app_util>)
+    list(APPEND AOM_EXAMPLE_TARGETS lightfield_tile_list_decoder)
+    list(APPEND AOM_APP_TARGETS lightfield_tile_list_decoder)
+  endif()
 
-if(ENABLE_EXAMPLES AND CONFIG_AV1_DECODER)
-  add_executable(lightfield_decoder "${AOM_ROOT}/examples/lightfield_decoder.c"
-                                    $<TARGET_OBJECTS:aom_common_app_util>
-                                    $<TARGET_OBJECTS:aom_decoder_app_util>)
-  list(APPEND AOM_EXAMPLE_TARGETS lightfield_decoder)
-  list(APPEND AOM_APP_TARGETS lightfield_decoder)
-endif()
+  if(ENABLE_EXAMPLES AND CONFIG_AV1_DECODER)
+    add_executable(lightfield_decoder "${AOM_ROOT}/examples/lightfield_decoder.c"
+                                      $<TARGET_OBJECTS:aom_common_app_util>
+                                      $<TARGET_OBJECTS:aom_decoder_app_util>)
+    list(APPEND AOM_EXAMPLE_TARGETS lightfield_decoder)
+    list(APPEND AOM_APP_TARGETS lightfield_decoder)
+  endif()
 
-if(ENABLE_EXAMPLES AND CONFIG_AV1_ENCODER AND CONFIG_AV1_DECODER)
-  add_executable(lightfield_bitstream_parsing
-                 "${AOM_ROOT}/examples/lightfield_bitstream_parsing.c"
-                 $<TARGET_OBJECTS:aom_common_app_util>
-                 $<TARGET_OBJECTS:aom_encoder_app_util>
-                 $<TARGET_OBJECTS:aom_decoder_app_util>)
-  list(APPEND AOM_EXAMPLE_TARGETS lightfield_bitstream_parsing)
-  list(APPEND AOM_APP_TARGETS lightfield_bitstream_parsing)
+  if(ENABLE_EXAMPLES AND CONFIG_AV1_ENCODER AND CONFIG_AV1_DECODER)
+    add_executable(lightfield_bitstream_parsing
+                   "${AOM_ROOT}/examples/lightfield_bitstream_parsing.c"
+                   $<TARGET_OBJECTS:aom_common_app_util>
+                   $<TARGET_OBJECTS:aom_encoder_app_util>
+                   $<TARGET_OBJECTS:aom_decoder_app_util>)
+    list(APPEND AOM_EXAMPLE_TARGETS lightfield_bitstream_parsing)
+    list(APPEND AOM_APP_TARGETS lightfield_bitstream_parsing)
+  endif()
 endif()
 
 foreach(aom_app ${AOM_APP_TARGETS})
-- 
2.50.1

